src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,332 @@
+/**
+ * @file gan.hpp
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_IMPL_HPP
+#define MLPACK_METHODS_ANN_GAN_IMPL_HPP
+
+#include "gan.hpp"
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/core/dists/gaussian_distribution.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::math;
+using namespace mlpack::distribution;
+
+namespace mlpack {
+namespace ann /** artifical neural network  */ {
+template<typename Generator, typename Discriminator, typename IntializerType>
+GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::GenerativeAdversarialNetwork(arma::mat& trainData,
+    IntializerType initializeRule,
+    Generator& generator,
+    Discriminator& discriminator,
+    size_t batchSize,
+    size_t noiseInSize,
+    size_t disIteration):
+    trainData(trainData),
+    initializeRule(initializeRule),
+    generator(generator),
+    discriminator(discriminator),
+    trainGenerator(false),
+    batchSize(batchSize),
+    noiseInSize(noiseInSize),
+    disIteration(2 * disIteration),
+    reset(false)
+{
+  discriminator.network.insert(discriminator.network.begin(),
+      new IdentityLayer<>());
+  trainGenerator = false;
+  iterations = 1;
+  offset = 0;
+  iterationDiscriminator = disIteration;
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Reset()
+{
+  std::cout << "Reset function" << std::endl;
+  size_t generatorWeights = 0;
+  size_t discriminatorWeights = 0;
+
+  for (size_t i = 0; i < generator.network.size(); ++i)
+    generatorWeights += boost::apply_visitor(weightSizeVisitor,
+        generator.network[i]);
+
+  assert(generatorWeights > 0);
+
+  for (size_t i = 0; i < discriminator.network.size(); ++i)
+    discriminatorWeights += boost::apply_visitor(weightSizeVisitor,
+        discriminator.network[i]);
+
+  assert(discriminatorWeights > 0);
+
+  parameter.set_size(generatorWeights + discriminatorWeights, 1);
```

Comment:
In order to get stable results, we have to be careful with the choice of architecture and parameter initialization. I think instead of using random weights we should test the ``GaussianInitialization`` or ``NguyenWidrowInitialization`. So instead of reimplementing the weight initialization, we should use the ``NetworkInitialization`` class, that way we are going to initialize each layer using the correct parameter space.

The only problem I see is that we would have to adjust the NetworkInitialization class so that we could call it with a predefined parameter matrix. So that once we call Initialize we don't resize the matrix.

```c++
void Initialize(const std::vector<LayerTypes>& network, arma::mat& parameter, const size_t parameterOffset = 0)
{
    if (parameter.is_empty())
    {
      // Determine the number of parameter/weights of the given network.
      size_t weights = 0;
      for (size_t i = 0; i < network.size(); ++i)
        weights += boost::apply_visitor(weightSizeVisitor, network[i]);
      parameter.set_size(weights, 1);
    }

    // Initialize the network layer by layer or the complete network.
    if (ann::InitTraits<InitializationRuleType>::UseLayer)
    {
      for (size_t i = 0, offset = parameterOffset; i < network.size(); ++i)
      {
        // Initialize the layer with the specified parameter/weight
        // initialization rule.
        const size_t weight = boost::apply_visitor(weightSizeVisitor,
            network[i]);
        arma::mat tmp = arma::mat(parameter.memptr() + offset,
            weight, 1, false, false);
        initializeRule.Initialize(tmp, tmp.n_elem, 1);

        // Increase the parameter/weight offset for the next layer.
        offset += weight;
      }
    }
    else
    {
      initializeRule.Initialize(parameter, parameter.n_elem, 1);
    }

    // Note: We can't merge the for loop into the for loop above because
    // WeightSetVisitor also sets the parameter/weights of the inner modules.
    // Inner Modules are held by the parent module e.g. the concat module can
    // hold various other modules.
    for (size_t i = 0, offset = parameterOffset; i < network.size(); ++i)
    {
      offset += boost::apply_visitor(WeightSetVisitor(std::move(parameter),
          offset), network[i]);

      boost::apply_visitor(resetVisitor, network[i]);
    }
}
```

Afterwards we can call the function like:

```c++
// Reset the network parameter with the given initialization rule.

parameter.set_size(generatorWeights + discriminatorWeights, 1);

NetworkInitialization<InitializationRuleType> networkInit(initializeRule);
networkInit.Initialize(generator, parameter);

networkInit.Initialize(discriminator, parameter, generatorWeights);
```